torch/csrc/cuda/shared/nvtx.cpp

```diff
@@ -12,15 +12,208 @@ namespace torch {
 namespace cuda {
 namespace shared {
 
+#ifndef USE_ROCM
+
+// registry idea from NVIDIA/nvtx-plugins  @jdekhtiar
+class DomainCategoryRegistry {
+ public:
+  DomainCategoryRegistry() = default;
+
+  ~DomainCategoryRegistry() {
+    for (auto& domain : domains) {
+      nvtxDomainDestroy(domain.second);
+    }
+  }
+
+  nvtxDomainHandle_t getDomain(const std::string& domain_name) {
+    if (const auto it = domains.find(domain_name); it != domains.end()) {
+      return it->second;
+    }
+
+    const auto domain_handle = nvtxDomainCreateA(domain_name.c_str());
+    domains[domain_name] = domain_handle;
+    return domain_handle;
+  }
+
+  uint64_t getCategory(
+      const std::string& category_name,
+      std::optional<std::string> domain_name) {
```

Comment:
std::optional is not trivially copyable since std::string is not. Use const ref instead (or move if appropriate)